tools/place_name_resolver/resolver.go

```diff
@@ -0,0 +1,285 @@
+package main
+
+import (
+	"cloud.google.com/go/storage"
+	"context"
+	"encoding/csv"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"googlemaps.github.io/maps"
+	"io"
+	"io/ioutil"
+	"log"
+	"os"
+	"sync"
+)
+
+var (
+	inCsvPath = flag.String("in_csv_path", "", "Input CSV file. Column names are supposed to be Place properties or "+
+		"literal 'Node' to represent a local ID. containedInPlace is always assumed to be a local reference.")
+	outCsvPath = flag.String("out_csv_path", "", "Same as input with additional column for DCID.")
+	mapsApiKey = flag.String("maps_api_key", "", "Key for accessing Geocoding Maps API.")
+)
+
+const (
+	batchSize          = 200
+	placeId2DcidBucket = "datcom-browser-prod.appspot.com"
+	placeId2DcidObject = "placeid2dcid.json"
+)
+
+//
+// PlaceId2Dcid Reader.
+//
+type PlaceId2Dcid interface {
+	Read() ([]byte, error)
+}
+
+type RealPlaceId2Dcid struct{}
+
+func (r *RealPlaceId2Dcid) Read() ([]byte, error) {
+	ctx := context.Background()
+	gcsCli, err := storage.NewClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+	fp, err := gcsCli.Bucket(placeId2DcidBucket).Object(placeId2DcidObject).NewReader(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer fp.Close()
+	return ioutil.ReadAll(fp)
+}
+
+//
+// Maps API Client.
+//
+type MapsClient interface {
+	Geocode(r *maps.GeocodingRequest) ([]maps.GeocodingResult, error)
+}
+
+type RealMapsClient struct {
+	Client *maps.Client
+}
+
+func (r *RealMapsClient) Geocode(req *maps.GeocodingRequest) ([]maps.GeocodingResult, error) {
+	return r.Client.Geocode(context.Background(), req)
+}
+
+//
+// Data Structure representing the CSV.
+//
+type tableInfo struct {
+	// CSV header.
+	header []string
+
+	// Ordered rows in CSV.
+	rows [][]string
+
+	// Expanded name used for geocoding, one per row.
+	extNames []string
+
+	// Map from node-id to row index.
+	node2row map[string]int
+
+	// Various indexes. -1 if missing.
+	nidIdx  int
+	cipIdx  int
```

Comment:
Maybe specify what nid (node id?) and cip (contained in place?) are?

Reply:
Improved comments.